### PR TITLE
Use headers for requestGet, retrieve, identity, find and describe calls

### PR DESCRIPTION
**Pull request by guanzhijie**
This is to follow up the previous pull request https://github.com/jsforce/jsforce/pull/494. I've made more improvements based on @stomita comments, and set headers for more operations.

### CODE REVIEW

**Comment by guanzhijie** [comment]
@stomita  can you please help check this pull request?

**Comment by asifalimd** [comment]
I am actually looking for a way to send headers as part of get request. The need is I need to set the   header( **"Sforce-Query-Options": "batchSize=500"**) when I am using Query/QueryMore. your code seems to be fine but there is one place where the GET is called directly not from requestGet. I think the headers needs to be set there too.
Because when I debug the code while making query/querymore it seems the code on this method is executing.
in Connection.prototype.request
```
// if request is simple string, regard it as url in GET method
  if (_.isString(request)) {
    request = { method: 'GET', url: request };
  } 
```

**Comment by guanzhijie** [comment]
@asifalimd I proposed that change in my previous pull request https://github.com/jsforce/jsforce/pull/494. See the conversation in that pull request, if I understand it correctly, I think @stomita prefers wrapping the headers in the request when making the call, rather than automatically adding the header into the request when the passed-in request is a url string in Connection.prototype.request.

**Comment by stomita** [comment]
@asifalimd  In the request layer, the http header information is passed in parameter and there is no options argument. The `request(url: string)` is the short version of `request({ method: 'GET', url: url })` when you don't have to set http headers. The CRUD/business methods layer it has options info, and accepts the `options.headers` to pass down to request layer.

@guanzhijie I've added a few comments please check.

File: lib/sobject.js [hunk 1]
```diff
@@ -530,12 +530,18 @@ ListView.prototype.results = function(callback) {
...
  */
-ListView.prototype.describe = function(callback) {
+ListView.prototype.describe = function(options, callback) {
+  if (typeof options === 'function') {
+    callback = options;
+    options = {};
+  }
   var url =  this._conn._baseUrl() + '/sobjects/' + this.type + '/listviews/' + this.id + '/describe';
-  return this._conn.request(url, callback);
+  return this._conn.request({ method: 'GET', url: url, headers: options.headers }, callback);
```
Comment (stomita) [inline lib/sobject.js]: Fails when neither `options` and `callback` are given, which are optional args.

File: lib/connection.js [hunk 2]
```diff
@@ -896,23 +897,29 @@ Connection.prototype.sobject = function(type) {
...
+    callback = options;
+    options = {};
+  }
   var self = this;
   var idUrl = this.userInfo && this.userInfo.url;
   return Promise.resolve(
     idUrl ?
     { identity: idUrl } :
-    this.request(this._baseUrl())
+    this.request({ method: 'GET', url: this._baseUrl(), headers: options.headers })
```
Comment (stomita) [inline lib/connection.js]: Fails when neither options and callback are given, which are optional args.

File: lib/connection.js [hunk 3]
```diff
@@ -333,7 +333,8 @@ Connection.prototype.request = function(request, options, callback) {
 Connection.prototype.requestGet = function(url, options, callback) {
   var request = {
     method: "GET",
-    url: url
+    url: url,
+    headers: options.headers
```
Comment (stomita) [inline lib/connection.js]: As the `options` argument is for `HttpApi` configuration, the http headers info is not to be here.

File: lib/sobject.js (hunk 1 above)
Comment (guanzhijie) [inline lib/sobject.js]: Thanks @stomita  for catching it. I added a line as "options = options || {};"

File: lib/connection.js (hunk 2 above)
Comment (guanzhijie) [inline lib/connection.js]: Thanks @stomita  for catching it. I added a line as "options = options || {};"

File: lib/connection.js (hunk 3 above)
Comment (guanzhijie) [inline lib/connection.js]: Got it. Removed.

**Comment by asifalimd** [comment]
@stomita, @guanzhijie , thanks for working on this. Adding the headers in options for HttpApi makes sense.

**Comment by guanzhijie** [comment]
@stomita  any more feedback for this pull request?

**Comment by tnoonan-salesforce** [comment]
@stomita Poke -

File: lib/connection.js [hunk 4]
```diff
@@ -896,23 +896,30 @@ Connection.prototype.sobject = function(type) {
...
-    this.request(this._baseUrl())
+    this.request({ method: 'GET', url: this._baseUrl(), headers: options.headers })
   ).then(function(res) {
     var url = res.identity;
     url += '?format=json&oauth_token=' + encodeURIComponent(self.accessToken);
     var transport = Transport.JsonpTransport.supported ?
       new Transport.JsonpTransport('callback') :
       undefined;
-    return self.requestGet(url, { transport: transport });
+    return self.requestGet(url, options, { transport: transport });
```
Comment (stomita) [inline lib/connection.js]: The interface of `requestGet` is `requestGet(url, options, callback)`, options to be included in 2nd arg.

**Pull request merged**